Add constant module

xldenis opened the pull request:
This is the continuation of https://github.com/llvm-hs/llvm-irbuilder/pull/21 but on the correct repo. 
I've added struct and array helpers. I don't know how frequently people use actual constant operations, we can add them as we go though.

xldenis committed:
Add constant module

cocreature commented:
Looks great, thanks again! The travis failure is legit: We use `NoImplicitPrelude` in `llvm-hs-pure`. Adding  `import LLVM.Prelude` should hopefully fix the errors.

xldenis commented:
Fixed. I also removed the unused `FlexibleInstances` pragma.

I'm not certain what to do about the ambiguous names? Should I fix the test case (by qualifying an import)? Or should `Constant` not be a default export?

xldenis commented:
Additionally, should I drop the `MonadIRBuilder` constraints on the methods entirely?  I have it there for consistency but really I'm just using `pure` everywhere.

cocreature commented:
Changing the constraint to `Applicative` seems like a good idea 👍

xldenis commented:
I've weakened the constraint to `Applicative` and squashed all the commits.

cocreature commented:
Thanks!
